docs/design/simplified_federation_deployment.md

```diff
@@ -0,0 +1,388 @@
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Background
+
+Kubernetes Cluster Federation has its own control plane which is similar
+to the control plane of a Kubernetes cluster. Like Kubernetes, a
+federation control plane is composed an API server and a controller
+manager that bundles all the core controllers. While that looks simple
+on paper, deploying a federation control plane comes with its own set of
+challenges. Deploying a federation control plane not just involves
+deploying a federation API server and a federation controller manager,
+but also other components that support the control plane. This includes:
+
+1. A persistent store to store the state of the federation - `etcd`.
+2. A persistent device to store `etcd` data.
+3. A secret store to store the federation API server credentials.
+
+And these components must be setup in such a way that they can
+appropriately coordinate with each other. While this is already starting
+to look complicated, this is only one part of the problem.
+
+Kubernetes Cluster Federation operates only on the clusters that it is
+responsible for. So once a federation control plane is deployed, for it
+to be useful, users should be able to register and deregister Kubernetes
+clusters with/from their federation. In order to register a cluster,
+users should supply the cluster's credentials and its API server
+endpoint.
+
+**Side note**: It is a mouthful to always prefix "federation" to each of
+its control plane components. But it is necessary to disambiguate these
+components from the almost identical components in a Kubernetes cluster.
+
+## Goals
+
+Today, we ship bash scripts to facilitate the federation control plane
+deployment process and the registration/deregistration process is mostly
+manual. Both these processes involve multiple coordinated steps and are
+error prone. For example, to register a Kubernetes cluster with
+federation users have to write a yaml file describing a cluster
+resource. They have to specify the cluster's endpoint in the right
+format. That field is too restrictive and hard to get it right. They
+also have to correctly specify where the credentials for the cluster is
+stored.
+
+Therefore, the goal here is to simplify both the federation control
+plane deployment experience and the cluster registration and
+deregistration process. Each of these processes will be reduced to a
+single command with additional flags for customization.
+
+
+## User Experience
+
+Kubernetes v1.4 shipped a new command line tool called
+[`kubeadm`](http://kubernetes.io/docs/getting-started-guides/kubeadm/).
+`kubeadm` simplifies Kubernetes control plane deployment. Federation
+control plane deployment builds on top of this tool. Cluster
+registration and deregistration will be facilitated through
+[`kubectl`](http://kubernetes.io/docs/user-guide/kubectl-overview/).
+
+
+#### Federation Control Plane Deployment Experience
+
+1. User provisions their cloud/compute resources: a VM or a physical
```

Comment:
Cloud resource provisioning is outside the scope of this document. Users could any of the methods they use today to provision resources. Even on GCE, there isn't just one. Why favor one or the other?